File: web/app/src/lib/api/reseller.ts
import { createHttpClient } from '@/lib/api/http'

const http = createHttpClient('reseller')

export type ResellerProfile = {
  id?: number
  user_id?: number
  name?: string
  contact_name?: string
  phone?: string
  notes?: string
  is_active?: boolean
  username?: string
  email?: string | null
  key_count?: number
  site_count?: number
  created_at?: string
  updated_at?: string
}

export type ResellerKey = {
  id?: number
  name?: string
  key_type?: string
  key_prefix?: string
  raw_key?: string
  viewable?: boolean
  is_active?: boolean
  site_count?: number
  last_used_at?: string | null
  created_at?: string
}

export type ResellerModelGroup = {
  id?: number
  code?: string
  name?: string
  model_count?: number
}

export type ResellerSite = {
  id?: number
  reseller_id?: number
  user_id?: number
  api_key_id?: number
  site_name?: string
  logo_url?: string
  domain?: string
  site_code?: string
  db_name?: string
  redis_db?: number
  app_port?: number
  nats_namespace?: string
  code_path?: string
  public_url?: string
  status?: string
  profit_ratio?: number
  smtp_host?: string
  smtp_port?: number
  smtp_user?: string
  smtp_from?: string
  last_error?: string
  created_at?: string
  updated_at?: string
}

export type ResellerBuildJob = {
  id?: number
  site_id?: number
  reseller_id?: number
  status?: string
  step?: string
  error?: string
  resources?: Record<string, unknown>
  started_at?: string
  finished_at?: string
  created_at?: string
  updated_at?: string
}

export type CreateResellerSitePayload = {
  api_key_id?: number
  site_name: string
  logo_url?: string
  domain?: string
  profit_ratio?: number
  smtp_host: string
  smtp_port?: number
  smtp_user: string
  smtp_password: string
  smtp_from: string
}

export const resellerAuthApi = {
  login: (payload: { username: string; password: string }) =>
    http.post<{ token: string; reseller?: ResellerProfile }>('/reseller/auth/login', payload),
}

export const resellerApi = {
  getProfile: () => http.get<ResellerProfile>('/reseller/profile'),
  getKeys: () =>
    http.get<{ keys?: ResellerKey[]; items?: ResellerKey[] } | ResellerKey[]>('/reseller/keys'),
  listModelGroups: () =>
    http.get<{ groups?: ResellerModelGroup[] }>('/user/model-groups'),
  createKey: (payload: { name: string; group_ids: number[] }) =>
    http.post<{ key?: string; note?: string }>('/reseller/keys', payload),
  getSites: () =>
    http.get<{ sites?: ResellerSite[]; items?: ResellerSite[] } | ResellerSite[]>('/reseller/sites'),
  createSite: (payload: CreateResellerSitePayload) =>
    http.post<{ site?: ResellerSite; job?: ResellerBuildJob }>('/reseller/sites', payload),
  getBuildProgress: (siteId: number) =>
    http.get<{ site?: ResellerSite; jobs?: ResellerBuildJob[] }>(`/reseller/sites/${siteId}/build-progress`),
}

File: internal/model/reseller.go
package model

import "time"

// Reseller is a platform-managed reseller account. The backing User is used
// for platform API keys so existing API authentication and billing stay intact.
type Reseller struct {
	ID          int64     `xorm:"pk autoincr 'id'" json:"id"`
	UserID      int64     `xorm:"notnull unique 'user_id'" json:"user_id"`
	Name        string    `xorm:"notnull default('') 'name'" json:"name"`
	ContactName string    `xorm:"notnull default('') 'contact_name'" json:"contact_name"`
	Phone       string    `xorm:"notnull default('') 'phone'" json:"phone"`
	Notes       string    `xorm:"text 'notes'" json:"notes,omitempty"`
	IsActive    bool      `xorm:"notnull default(true) 'is_active'" json:"is_active"`
	CreatedAt   time.Time `xorm:"created 'created_at'" json:"created_at"`
	UpdatedAt   time.Time `xorm:"updated 'updated_at'" json:"updated_at"`
}

func (*Reseller) TableName() string { return "resellers" }

type ResellerSite struct {
	ID            int64     `xorm:"pk autoincr 'id'" json:"id"`
	ResellerID    int64     `xorm:"notnull index 'reseller_id'" json:"reseller_id"`
	UserID        int64     `xorm:"notnull index 'user_id'" json:"user_id"`
	APIKeyID      int64     `xorm:"notnull index 'api_key_id'" json:"api_key_id"`
	SiteName      string    `xorm:"notnull 'site_name'" json:"site_name"`
	LogoURL       string    `xorm:"notnull default('') 'logo_url'" json:"logo_url"`
	Domain        string    `xorm:"notnull default('') 'domain'" json:"domain"`
	SiteCode      string    `xorm:"notnull unique 'site_code'" json:"site_code"`
	DBName        string    `xorm:"notnull unique 'db_name'" json:"db_name"`
	RedisDB       int       `xorm:"notnull unique 'redis_db'" json:"redis_db"`
	AppPort       int       `xorm:"notnull unique 'app_port'" json:"app_port"`
	NATSNamespace string    `xorm:"notnull unique 'nats_namespace'" json:"nats_namespace"`
	CodePath      string    `xorm:"notnull 'code_path'" json:"code_path"`
	PublicURL     string    `xorm:"notnull default('') 'public_url'" json:"public_url"`
	Status        string    `xorm:"notnull default('pending') index 'status'" json:"status"`
	ProfitRatio   float64   `xorm:"notnull default(1.7) 'profit_ratio'" json:"profit_ratio"`
	SMTPHost      string    `xorm:"notnull default('') 'smtp_host'" json:"smtp_host"`
	SMTPPort      int       `xorm:"notnull default(465) 'smtp_port'" json:"smtp_port"`
	SMTPUser      string    `xorm:"notnull default('') 'smtp_user'" json:"smtp_user"`
	SMTPPassword  string    `xorm:"text 'smtp_password'" json:"-"`
	SMTPFrom      string    `xorm:"notnull default('') 'smtp_from'" json:"smtp_from"`
	LastError     string    `xorm:"text 'last_error'" json:"last_error,omitempty"`
	CreatedAt     time.Time `xorm:"created 'created_at'" json:"created_at"`
	UpdatedAt     time.Time `xorm:"updated 'updated_at'" json:"updated_at"`
}

func (*ResellerSite) TableName() string { return "reseller_sites" }

type ResellerSiteBuildJob struct {
	ID         int64      `xorm:"pk autoincr 'id'" json:"id"`
	SiteID     int64      `xorm:"notnull index 'site_id'" json:"site_id"`
	ResellerID int64      `xorm:"notnull index 'reseller_id'" json:"reseller_id"`
	Status     string     `xorm:"notnull default('pending') index 'status'" json:"status"`
	Step       string     `xorm:"notnull default('pending') 'step'" json:"step"`
	Error      string     `xorm:"text 'error'" json:"error,omitempty"`
	Resources  JSON       `xorm:"jsonb 'resources'" json:"resources"`
	StartedAt  *time.Time `xorm:"'started_at' null" json:"started_at,omitempty"`
	FinishedAt *time.Time `xorm:"'finished_at' null" json:"finished_at,omitempty"`
	CreatedAt  time.Time  `xorm:"created 'created_at'" json:"created_at"`
	UpdatedAt  time.Time  `xorm:"updated 'updated_at'" json:"updated_at"`
}

func (*ResellerSiteBuildJob) TableName() string { return "reseller_site_build_jobs" }

type ResellerSiteKeyBinding struct {
	ID         int64     `xorm:"pk autoincr 'id'" json:"id"`
	ResellerID int64     `xorm:"notnull index 'reseller_id'" json:"reseller_id"`
	SiteID     int64     `xorm:"notnull index 'site_id'" json:"site_id"`
	APIKeyID   int64     `xorm:"notnull index 'api_key_id'" json:"api_key_id"`
	IsActive   bool      `xorm:"notnull default(true) 'is_active'" json:"is_active"`
	CreatedAt  time.Time `xorm:"created 'created_at'" json:"created_at"`
}

func (*ResellerSiteKeyBinding) TableName() string { return "reseller_site_key_bindings" }
